File: solve.py
#!/usr/bin/python2

# Credits:
# Algorithm design and implementation by
# Marcin Sulikowski (@marcinsulikowski)

from __future__ import print_function
from __future__ import division
import codecs
import math
import sys
import traceback


class DictionaryNode(object):
	__slots__ = ['finished', '_next']

	def __init__(self):
		self.finished = False
		self._next = {}

	def add_suffix(self, suffix):
		if not suffix:
			self.finished = True
		else:
			char = suffix[0]
			if not char in self._next:
				self._next[char] = DictionaryNode()
			self._next[char].add_suffix(suffix[1:])

	def get(self, char):
		return self._next.get(char, None)

	def print_all(self, prefix=""):
		if self.finished:
			print(prefix)
		for char in self._next:
			self._next[char].print_all(prefix + char)


class Board(object):
	class Cube(object):
		def __init__(self, letter):
			self.letter = letter
			self.visited = False
			self.neighbors = []

	def __init__(self, letters):
		size = int(math.sqrt(len(letters)))
		assert size * size == len(letters)
		self._size = size
		self._cubes = [self.Cube(l) for l in letters]
		deltas = [(-1,-1),(-1,0),(-1,1),(0,-1),(0,1),(1,-1),(1,0),(1,1)]
		for x in range(size):
			for y in range(size):
				for (dx, dy) in deltas:
					nx, ny = x + dx, y + dy
					if nx >= 0 and nx < size and ny >= 0 and ny < size:
						self.get_cube(x, y).neighbors.append(self.get_cube(nx, ny))

	def get_cube(self, x, y):
		return self._cubes[y * self._size + x]

	def solve(self, dictionary):
		result = set()
		for cube in self._cubes:
			self._solve_recursive(result, "", cube, dictionary)
		return sorted(result, key=len)

	def _solve_recursive(self, result, prefix, cube, dict_node):
		next_node = dict_node.get(cube.letter)
		if next_node is None:
			return
		cube.visited = True
		new_prefix = prefix + cube.letter
		if next_node.finished and len(new_prefix) >= 3:
			result.add(new_prefix)
		for neighbor in cube.neighbors:
			if not neighbor.visited:
				self._solve_recursive(result, new_prefix, neighbor, next_node)
		cube.visited = False


def main():
	dictionary_root = DictionaryNode()
	with codecs.open(sys.argv[1], encoding="latin2") as dict_file:
		for line in dict_file:
			dictionary_root.add_suffix(line.strip())
	print('[ OK ] Ready', file=sys.stderr)
	letters = ''
	while True:
		line = sys.stdin.readline()
		if not line:
			break
		line = line.strip()
		if letters and not line:
			print('[ OK ] Solving', file=sys.stderr)
			try:
				board = Board(unicode(letters, encoding='utf-8'))
				for word in board.solve(dictionary_root):
					print(u'({}) {}'.format(len(word), word))
				print('[ OK ] Solved', file=sys.stderr)
			except Exception as e:
				print('[FAIL] {}:\n{}'.format(e, traceback.format_exc()))
			letters = ''
		else:
			letters = letters + line


if __name__ == '__main__':
	main()

# vim: ft=python

File: solve.lua
#! /usr/bin/env lua

function ziter( a, i )
	i = i + 1
	local v = a[ i ]
	if v then
		return i, v
	end
end

function zpairs( a )
	return ziter, a, -1
end

function hs( h )
	local count = 0
	for k in pairs( h ) do
		count = count + 1
	end
	return count
end

DictionaryNode = {
	_finished = false,
}

function DictionaryNode:new()
	local o = { _next = {} }
	setmetatable( o, self )
	self.__index = self
	return o
end

function DictionaryNode:add_suffix( suffix_ )
	if suffix_:len() == 0 then
		self._finished = true
	else
		local char = suffix_:sub( 1, 1 )
		if self._next[ char ] == nil then
			self._next[ char ] = DictionaryNode:new()
		end
		self._next[ char ]:add_suffix( suffix_:sub( 2 ) )
	end
end

function DictionaryNode:get( char_ )
	return self._next[ char_ ]
end

function DictionaryNode:print_all( prefix_ )
	if prefix_ == nil then
		prefix_ = ""
	end
	if self._finished then
		print( prefix_ )
	end
	for char in pairs( self._next ) do
		self._next[ char ]:print_all( prefix_ .. char )
	end
end

Cube = {
	_visited = false
}

function Cube:new( letter_ )
	local o = { _neighbors = {}, _letter = letter_ }
	setmetatable( o, self )
	self.__index = self
	return o
end

Board = {
	_size = 0
}

function Board:get_cube( x_, y_ )
	return self._cubes[y_ * self._size + x_]
end

function Board:new( letters_ )
	local o = { _cubes = {} }
	setmetatable( o, self )
	self.__index = self
	local len = math.sqrt( letters_:len() )
	if len * len ~= letters_:len() then
		return nil
	end
	o._size = len
	for i = 1, #letters_ do
		o._cubes[i - 1] = Cube:new( letters_:sub( i, i ) )
	end
	local deltas = { { -1, -1 }, { -1, 0 }, { -1, 1 }, { 0, -1 }, { 0, 1 }, { 1, -1 }, { 1, 0 }, { 1, 1 } }
	for x = 0, len - 1 do
		for y = 0, len - 1 do
			for _, d in pairs( deltas ) do
				local nx = x + d[1]
				local ny = y + d[2]
				if ( nx >= 0 ) and ( nx < len ) and ( ny >= 0 ) and ( ny < len ) then
					c = o:get_cube( x, y )
					c._neighbors[#(c._neighbors) + 1] = o:get_cube( nx, ny )
				end
			end
		end
	end
	return o
end

function Board:solve( dictionary_ )
	local result = {}
	for _, cube in zpairs( self._cubes ) do
		self:solve_recursive( result, "", cube, dictionary_ )
	end
	local sorted = {}
	for word in pairs( result ) do
		sorted[#sorted + 1] = word
	end
	table.sort(
		sorted,
		function( a, b )
			return a:len() < b:len()
		end
	)
	return sorted
end

function Board:solve_recursive( result, prefix_, cube_, dictNode_ )
	local nextNode = dictNode_:get( cube_._letter )
	if nextNode == nil then
		return
	end
	cube_._visited = true
	local newPrefix = prefix_ .. cube_._letter
	if nextNode._finished and ( newPrefix:len() >= 3 ) then
		result[ newPrefix ] = true
	end
	for _, neighbor in pairs( cube_._neighbors ) do
		if not neighbor._visited then
			self:solve_recursive( result, newPrefix, neighbor, nextNode )
		end
	end
	cube_._visited = false
end

function main( argv_ )
	local dictionaryRoot = DictionaryNode:new()
	local dictFile = io.open( argv_[1], "r" )
	for line in dictFile:lines() do
		dictionaryRoot:add_suffix( line )
	end
	dictFile:close()
--	dictionaryRoot:print_all()
	print( "[ OK ] Ready" )
	local letters = ""
	while ( true ) do
		line = io.read()
		if line == nil then
			break;
		end
		if ( letters:len() > 0 ) and ( line:len() == 0 ) then
			print( "[ OK ] Solving" )
			board = Board:new( letters )
			if board ~= nil then
				for _, word in pairs( board:solve( dictionaryRoot ) ) do
					print( "(" .. word:len() .. ") " .. word )
				end
				print( "[ OK ] Solved" )
			else
				print( "Bad board definition!" )
			end
			letters = "";
		else
			letters = letters .. line
		end
	end
	return ( 0 );
end

main( arg )

-- vim: set ft=lua:
